### PR TITLE
Simplify CCR connection retry logic.

### PR DESCRIPTION
**Pull request by idegtiarenko**
This change attempts to simplify retry logic by removing unnecessary nesting.

### CODE REVIEW

**Review by DaveCTurner — COMMENTED** [review]
Seems reasonable; could we also remove the `if (seedNodesSuppliers.hasNext()) {` at the top of the method and move that to the original caller, since it's done at the recursive call site too?

**Comment by idegtiarenko** [comment]
> could we also remove the if (seedNodesSuppliers.hasNext()) { at the top of the method

Prior code guards us from entering this method when somebody removes the remote connections or sets the seeds to empty. I would suggest adding an assertion to enforce that state if we would like to remove the condition.

File: server/src/main/java/org/elasticsearch/transport/SniffConnectionStrategy.java [hunk 1]
```diff
@@ -228,126 +228,111 @@ private void collectRemoteNodes(Iterator<Supplier<DiscoveryNode>> seedNodesSuppl
...
-                    if (seedNodesSuppliers.hasNext()) {
-                        logger.debug(
-                            () -> format("fetching nodes from external cluster [%s] failed moving to next seed node", clusterAlias),
-                            e
-                        );
-                        collectRemoteNodes(seedNodesSuppliers, listener);
-                        return;
-                    }
-                }
+        assert seedNodesSuppliers.hasNext() : "Must have non empty seed nodes to connect";
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/SniffConnectionStrategy.java:231]: Hmm now that I look more closely, I'm not sure this does always hold. We also get here via the two implementations of `AbstractSniffResponseHandler` which do not look like they guard against running out of nodes properly.

Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/transport/SniffConnectionStrategy.java:231]: Yeah, this is in-fact failed in the last build. Seems we have to keep that condition.

**Review by DaveCTurner — APPROVED** [review]
LGTM as of [95dd2b4](https://github.com/elastic/elasticsearch/pull/96709/commits/95dd2b49d01c8477d9de60d7bdfb2b57bdf653ae) - this is a sensible change as it stands, but I'd still like a follow-up to avoid a bare `NoSeedNodeLeftException` when we have more detailed failure info for the last failure.

**Pull request merged**